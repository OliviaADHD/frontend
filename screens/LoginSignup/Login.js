import React, {useState, useEffect} from "react";
import { StatusBar } from "expo-status-bar";
import { Formik } from "formik";
import {Ionicons} from '@expo/vector-icons';
import { Alert, View } from "react-native";
import RootStack from "../../navigators/RootStack";
import* as AuthSession from 'expo-auth-session';
import * as Google from 'expo-auth-session/providers/google';
import * as Facebook from 'expo-auth-session/providers/facebook';
import { ResponseType } from 'expo-auth-session';
import * as WebBrowser from 'expo-web-browser';

import {
  StyledContainer,
  InnerContainer,
  PageLogo,
  StyledFormArea,
  StyledTextInput,
  RightIcon,
  StyledButton,
  ButtonText,
  ForgotPasswordText,
  ForgotPassword,
  Or,
  IconContainer,
  IconLogo,
  EachIconContainer,
  ExtraView,
  ExtraText,
  TextLink,
  TextLinkContent,
} from '../../components/styles';


WebBrowser.maybeCompleteAuthSession();

const Login = ({navigation}) => {
    const [hidePassword, setHidePassword] = useState(true);
    const [GoogleLogin, setGoogleLogin] = useState("texxt");
    const [request, response, promptAsync] = Google.useAuthRequest({
        expoClientId: '51546200734-nm24i67drlpn5dkcnaj4ckta6k2cnfff.apps.googleusercontent.com',
        webClientId: '51546200734-qv54r4ur316rk4ll8lb37esgstngnr4i.apps.googleusercontent.com',
        ClientSecret: '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',
      });

      const [requestFb, responseFb, promptAsyncFb] = Facebook.useAuthRequest({
        expoClientId: '901148574169051',
        responseType: ResponseType.Code,
      });
      

    useEffect(() => {
        if (response?.type === 'success'){
            console.log("success logging in with google! Continue on?");
            const {authentication } = response;
        }

    }, [response]);

    useEffect(() =>{
        if (responseFb?.type === 'success'){
            console.log("successfulley logged in with fb!");
            const {code} = responseFb.params;            
        }
    }, [responseFb]);
    
    return(
        <StyledContainer>
            <StatusBar style="dark"/>
            <InnerContainer>
                <PageLogo source={require('../../assets/images/logo.png')} />
                <Formik
                    initialValues={{email: '', password: ''}}
                    onSubmit={(values) => {
                        console.log(values['email']);
                        if (values['email']===""){console.log('empty')}
                        console.log(values);
                    }}
                >{({handleChange, handleBlur, handleSubmit, values})=> (<StyledFormArea>
                    <MyTextInput 
                        placeholder="Email"
                        onChangeText={handleChange('email')}
                        onBlur={handleBlur('email')}
                        value={values.email}
                        keyboardType="email-address"
                    />

                    <MyTextInput 
                        placeholder="Password"
                        onChangeText={handleChange('password')}
                        onBlur={handleBlur('password')}
                        value={values.password}
                        secureTextEntry={hidePassword}
                        isPassword={true}
                        hidePassword={hidePassword}
                        setHidePassword={setHidePassword}
                    />
                    <ForgotPassword>
                        <ForgotPasswordText>Forgot Password?</ForgotPasswordText>
                    </ForgotPassword>
                    <StyledButton onPress={handleSubmit}>
                        <ButtonText>Login</ButtonText>
                    </StyledButton>
                    <Or>Or</Or>
                    <IconContainer>
                        <EachIconContainer onPress={async()=>{
                            console.log("Trying to log in with google");
                            promptAsync();
                            console.log("success?")

                        }}>
                            <IconLogo source={require('../../assets/images/google.png')} />
                        </EachIconContainer>
                        <EachIconContainer onPress={() => {
                            console.log("trying to log in with fb");
                            promptAsyncFb();
                            console.log("success with fb?");}} >
                            <IconLogo source={require('../../assets/images/facebook.png')} />
                        </EachIconContainer>
                        <EachIconContainer onPress={handleSubmit}>
                            <IconLogo  source={require('../../assets/images/apple.png')} />
                        </EachIconContainer>
                    </IconContainer>
                    <ExtraView>
                        <ExtraText>Would You Like To Join Us? </ExtraText>
                        <TextLink onPress = {() => navigation.navigate("Signup")}>
                            <TextLinkContent>Signup</TextLinkContent>
                        </TextLink>
                    </ExtraView>
                </StyledFormArea>)}
                </Formik>
            </InnerContainer>
        </StyledContainer>   
    );
};



const MyTextInput = ({isPassword, icon, hidePassword, setHidePassword, ...props}) => {
    return (
        <View>
            <StyledTextInput {...props} />
            {isPassword && (
                <RightIcon onPress={() => setHidePassword(!hidePassword)}>
                    <Ionicons name={hidePassword ? 'md-eye-off' : 'md-eye'} size={19}/>
                </RightIcon>
            )}
        </View>
    )
}

export default Login;
