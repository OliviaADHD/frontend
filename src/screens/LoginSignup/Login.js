import React, {useState, useEffect} from "react";
import { StatusBar } from "expo-status-bar";
import { Formik } from "formik";
import {Ionicons} from '@expo/vector-icons';
import { StyleSheet, ActivityIndicator, View } from "react-native";
import { useDispatch, useSelector } from "react-redux";
import RootStack from "../../navigators/RootStack";
import* as AuthSession from 'expo-auth-session';
import * as Facebook from 'expo-auth-session/providers/facebook';
import { ResponseType } from 'expo-auth-session';
import * as WebBrowser from 'expo-web-browser';
import { signIn, beforeSignIn, signInGoogle } from '../../redux/actions/user/login'
import {
  PageLogo,
  StyledFormArea,
  StyledTextInput,
  RightIcon,
  StyledButton,
  ButtonText,
  ForgotPasswordText,
  ForgotPassword,
  Or,
  IconContainer,
  IconLogo,
  EachIconContainer,
  ExtraView,
  ExtraText,
  TextLink,
  TextLinkContent,
  ErrorText,
  ErrorMessage,
} from '../../css/LoginSignup/login';

import {  
    StyledContainer,
    InnerContainer,
    Loading
} from '../../css/general/style';

import * as yup from 'yup';

import * as Google from 'expo-auth-session/providers/google';
import axios from "axios";
import {link} from '../../redux/config/config';


const signInValidationSchema = yup.object().shape({
    email: yup
      .string().label('Email')
      .email("Please enter valid email")
      .required('Email is required'),
    password: yup
      .string().label('Password')
  })


//WebBrowser.maybeCompleteAuthSession();

const Login = ({navigation}) => {
    const dispatch = useDispatch();
    const [hidePassword, setHidePassword] = useState(true);
    const [loading, setLoading] = useState(false);
    const userData = useSelector(state => state.userName);
    const loginState = useSelector(state => state.loginInfo);
    const networkError = useSelector(state => state.networkAvailability);

    //Google login
    const [googleError, setGoogleError] = useState(false);
    const [googleClicked, setGoogleClicked] = useState(false);
    const [requestGoogle, responseGoogle, promptAsyncGoogle] = Google.useAuthRequest({
        expoClientId: '51546200734-nm24i67drlpn5dkcnaj4ckta6k2cnfff.apps.googleusercontent.com',
        iosClientId: '51546200734-nm24i67drlpn5dkcnaj4ckta6k2cnfff.apps.googleusercontent.com',
        webClientId: '51546200734-qv54r4ur316rk4ll8lb37esgstngnr4i.apps.googleusercontent.com',
        ClientSecret: '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',
      });
    
    const makeGoogleLogin = async(token) => {
        dispatch(beforeSignIn())
        .then(() => dispatch(signInGoogle(token)))
        .then((resp) => {
            if (resp.success === true){
                setLoading(false);
                setGoogleClicked(false);
                
                if (resp.firstTime){

                    navigation.reset({
                        index: 0,
                        routes: [{ name: 'Welcome_Post_Signup' }]});
                    }
                else {
                    navigation.reset({
                        index: 0,
                        routes: [{ name: 'Home' }]});
                }

            } else {
                setGoogleError(true);
                setLoading(false);
                setGoogleClicked(false);
            }

        })
        .catch(err => console.log(error, err));
    } 

    useEffect(() => {
        if ((responseGoogle?.type === 'success') && (googleClicked === true)){
            setGoogleClicked(false);
            const {authentication: { accessToken } } = responseGoogle;
            //getGoogleData(accessToken);
            makeGoogleLogin(accessToken);
            

        } else {setLoading(false);}

    }, [responseGoogle]);





    /*const [GoogleLogin, setGoogleLogin] = useState("texxt");
    const [request, response, promptAsync] = Google.useAuthRequest({
        expoClientId: '51546200734-nm24i67drlpn5dkcnaj4ckta6k2cnfff.apps.googleusercontent.com',
        iosClientId: '51546200734-nm24i67drlpn5dkcnaj4ckta6k2cnfff.apps.googleusercontent.com',
        webClientId: '51546200734-qv54r4ur316rk4ll8lb37esgstngnr4i.apps.googleusercontent.com',
        ClientSecret: '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',
      });

      const [requestFb, responseFb, promptAsyncFb] = Facebook.useAuthRequest({
        expoClientId: '901148574169051',
        responseType: ResponseType.Code,
      });
      
    
    useEffect(() => {
        if (response?.type === 'success'){
            console.log("success logging in with google! Continue on?");
            const {authentication } = response;
        }

    }, [response]);

    useEffect(() =>{
        if (responseFb?.type === 'success'){
            console.log("successfulley logged in with fb!");
            const {code} = responseFb.params;            
        }
    }, [responseFb]);
    */
    
    
    

    /*
    useEffect(() => {
        if (loginState.message.passed){
            if (userData.firstTime) {
                navigation.reset('Welcome_Post_Signup');
            } else {
                navigation.reset('Home', {
                    name: userData.Name,
                    id: userData.ID,
                    firstTime: userData.firstTime,
                });
            }
        }
    });*/

    return(
            <StyledContainer>
                <StatusBar style="dark"/>
                <InnerContainer>
                    <PageLogo source={require('../../../assets/images/logo.png')} />
                    <Formik
                        initialValues={{email: '', password: ''}}
                        validationSchema={signInValidationSchema}
                        onSubmit={(values) => {
                            setLoading(true);
                            console.warn("dum");
                            dispatch(beforeSignIn())
                            .then(() => dispatch(signIn(values)))
                            .then((resp) => {
                                if (resp.success === true){
                                    console.log('successfully logged in!', resp.firstTime);
                                    setLoading(false);                                    
                                    if (resp.firstTime){
                                        navigation.reset({
                                            index: 0,
                                            routes: [{ name: 'Welcome_Post_Signup' }]});
                                        }
                                    else {
                                        navigation.reset({
                                            index: 0,
                                            routes: [{ name: 'Home' }]});
                                    }
                                } else {
                                    setLoading(false);
                                }
                            });
                        }}
                        
                    
                    >{({handleChange, handleBlur, handleSubmit, values})=> (<StyledFormArea>
                        <MyTextInput 
                            placeholder="Email"
                            onChangeText={handleChange('email')}
                            onBlur={handleBlur('email')}
                            value={values.email}
                            keyboardType="email-address"
                        />

                        <MyTextInput 
                            placeholder="Password"
                            onChangeText={handleChange('password')}
                            onBlur={handleBlur('password')}
                            value={values.password}
                            secureTextEntry={hidePassword}
                            isPassword={true}
                            hidePassword={hidePassword}
                            setHidePassword={setHidePassword}
                        />
                        <ForgotPassword onPress = {() =>  navigation.navigate("ResetPassword")}>
                            <ForgotPasswordText>Forgot Password?</ForgotPasswordText>
                        </ForgotPassword>
                        <StyledButton onPress={handleSubmit}>
                            <ButtonText>Login</ButtonText>
                        </StyledButton>
                        {networkError.error && 
                            <ErrorMessage>
                                <ErrorText>No network connection. Please try again later.</ErrorText>
                            </ErrorMessage>
                        }
                        { loginState.message.error &&
                            <ErrorMessage>
                                <ErrorText>{loginState.message.errorMessage}</ErrorText>
                            </ErrorMessage> 
                        }
                        {googleError && 
                            <ErrorMessage>
                                <ErrorText>Error signing in with google.</ErrorText>
                            </ErrorMessage>
                        }

                        <Or>Or</Or>
                        <IconContainer>
                            <EachIconContainer onPress={async()=>{
                                console.log("Trying to log in with google");
                                setLoading(true);
                                setGoogleClicked(true);
                                promptAsyncGoogle();
                                console.log("success?")

                            }}>
                                <IconLogo source={require('../../../assets/images/google.png')} />
                            </EachIconContainer>
                            <EachIconContainer onPress={() => {
                                console.log("trying to log in with fb");
                                promptAsyncFb();
                                console.log("success with fb?");}} >
                                <IconLogo source={require('../../../assets/images/facebook.png')} />
                            </EachIconContainer>
                            <EachIconContainer onPress={handleSubmit}>
                                <IconLogo  source={require('../../../assets/images/apple.png')} />
                            </EachIconContainer>
                        </IconContainer>
                        <ExtraView>
                            <ExtraText>Would You Like To Join Us? </ExtraText>
                            <TextLink onPress = {() =>  navigation.navigate("Signup")}>
                                <TextLinkContent>Signup</TextLinkContent>
                            </TextLink>
                        </ExtraView>
                    </StyledFormArea>)}
                    </Formik>
                </InnerContainer>
                {loading &&
                    <Loading>
                        <ActivityIndicator size="large" color="#694398"/>
                    </Loading>
                }
            </StyledContainer>   
        );
    
    }

const MyTextInput = ({isPassword, icon, hidePassword, setHidePassword, ...props}) => {
    return (
        <View>
            <StyledTextInput {...props} />
            {isPassword && (
                <RightIcon onPress={() => setHidePassword(!hidePassword)}>
                    <Ionicons name={hidePassword ? 'md-eye-off' : 'md-eye'} size={19}/>
                </RightIcon>
            )}
        </View>
    )
}


export default Login;
