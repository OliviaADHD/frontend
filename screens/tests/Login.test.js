import React from 'react';
import {render, fireEvent, act} from '@testing-library/react-native';
import renderer from 'react-test-renderer';

import Login from '../Login';



const mockNavigation = {
    navigate: jest.fn(),
};

it('renders correctly', () => {
    const tree = renderer.create(<Login />).toJSON();
    expect(tree).toMatchSnapshot();
  });

