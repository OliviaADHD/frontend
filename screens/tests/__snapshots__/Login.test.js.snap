// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly 1`] = `
<View
  style={
    Array [
      Object {
        "backgroundColor": "#ffffff",
        "flexBasis": 0,
        "flexGrow": 1,
        "flexShrink": 1,
        "marginTop": 44,
      },
    ]
  }
>
  <View
    style={
      Array [
        Object {
          "alignItems": "center",
          "flexBasis": 0,
          "flexGrow": 1,
          "flexShrink": 1,
          "width": "100%",
        },
      ]
    }
  >
    <Image
      source={1}
      style={
        Array [
          Object {
            "height": "20%",
            "marginBottom": "10%",
            "marginTop": "5%",
            "resizeMode": "contain",
            "width": "100%",
          },
        ]
      }
    />
    <View
      style={
        Array [
          Object {
            "height": "75%",
            "width": "90%",
          },
        ]
      }
    >
      <View>
        <TextInput
          allowFontScaling={true}
          keyboardType="email-address"
          onBlur={[Function]}
          onChangeText={[Function]}
          placeholder="Email"
          rejectResponderTermination={true}
          style={
            Array [
              Object {
                "backgroundColor": "#ffffff",
                "borderBottomWidth": 1,
                "fontSize": 17,
                "height": 40,
                "marginBottom": "3%",
                "marginLeft": "4%",
                "marginRight": "4%",
                "marginTop": "3%",
                "textAlign": "left",
              },
            ]
          }
          underlineColorAndroid="transparent"
          value=""
        />
      </View>
      <View>
        <TextInput
          allowFontScaling={true}
          onBlur={[Function]}
          onChangeText={[Function]}
          placeholder="Password"
          rejectResponderTermination={true}
          secureTextEntry={true}
          style={
            Array [
              Object {
                "backgroundColor": "#ffffff",
                "borderBottomWidth": 1,
                "fontSize": 17,
                "height": 40,
                "marginBottom": "3%",
                "marginLeft": "4%",
                "marginRight": "4%",
                "marginTop": "3%",
                "textAlign": "left",
              },
            ]
          }
          underlineColorAndroid="transparent"
          value=""
        />
        <View
          accessible={true}
          focusable={true}
          onClick={[Function]}
          onResponderGrant={[Function]}
          onResponderMove={[Function]}
          onResponderRelease={[Function]}
          onResponderTerminate={[Function]}
          onResponderTerminationRequest={[Function]}
          onStartShouldSetResponder={[Function]}
          style={
            Object {
              "left": "90%",
              "opacity": 1,
              "right": "20%",
              "top": "-50%",
              "zIndex": 1,
            }
          }
        >
          <Text />
        </View>
      </View>
      <View
        accessible={true}
        focusable={false}
        onClick={[Function]}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          Object {
            "backgroundColor": "#ffffff",
            "height": "15%",
            "opacity": 1,
          }
        }
      >
        <Text
          style={
            Array [
              Object {
                "fontSize": 15,
                "height": "100%",
                "marginBottom": "0.5%",
                "marginLeft": "4%",
                "marginTop": "0.5%",
                "width": 127,
              },
            ]
          }
        >
          Forgot Password?
        </Text>
      </View>
      <View
        accessible={true}
        focusable={true}
        onClick={[Function]}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          Object {
            "backgroundColor": "#694398",
            "borderRadius": 5,
            "height": "8%",
            "justifyContent": "center",
            "marginBottom": "0%",
            "marginLeft": "11%",
            "marginRight": "11%",
            "marginTop": "0%",
            "opacity": 1,
            "paddingHorizontal": "1%",
            "paddingVertical": "1%",
          }
        }
      >
        <Text
          style={
            Array [
              Object {
                "color": "#ffffff",
                "fontSize": 20,
                "textAlign": "center",
              },
            ]
          }
        >
          Login
        </Text>
      </View>
      <Text
        style={
          Array [
            Object {
              "fontSize": 15,
              "height": "5%",
              "marginLeft": "48%",
              "marginTop": "4%",
              "textAlign": "center",
              "width": 18,
            },
          ]
        }
      >
        Or
      </Text>
      <View
        style={
          Array [
            Object {
              "flexDirection": "row",
              "height": "10%",
              "justifyContent": "space-between",
              "marginBottom": "5%",
              "marginLeft": "22%",
              "marginTop": "2%",
              "width": "50%",
            },
          ]
        }
      >
        <View
          accessible={true}
          focusable={true}
          onClick={[Function]}
          onResponderGrant={[Function]}
          onResponderMove={[Function]}
          onResponderRelease={[Function]}
          onResponderTerminate={[Function]}
          onResponderTerminationRequest={[Function]}
          onStartShouldSetResponder={[Function]}
          style={
            Object {
              "height": "100%",
              "opacity": 1,
              "width": "30%",
            }
          }
        >
          <Image
            source={1}
            style={
              Array [
                Object {
                  "height": "100%",
                  "resizeMode": "contain",
                },
              ]
            }
          />
        </View>
        <View
          accessible={true}
          focusable={true}
          onClick={[Function]}
          onResponderGrant={[Function]}
          onResponderMove={[Function]}
          onResponderRelease={[Function]}
          onResponderTerminate={[Function]}
          onResponderTerminationRequest={[Function]}
          onStartShouldSetResponder={[Function]}
          style={
            Object {
              "height": "100%",
              "opacity": 1,
              "width": "30%",
            }
          }
        >
          <Image
            source={1}
            style={
              Array [
                Object {
                  "height": "100%",
                  "resizeMode": "contain",
                },
              ]
            }
          />
        </View>
        <View
          accessible={true}
          focusable={true}
          onClick={[Function]}
          onResponderGrant={[Function]}
          onResponderMove={[Function]}
          onResponderRelease={[Function]}
          onResponderTerminate={[Function]}
          onResponderTerminationRequest={[Function]}
          onStartShouldSetResponder={[Function]}
          style={
            Object {
              "height": "100%",
              "opacity": 1,
              "width": "30%",
            }
          }
        >
          <Image
            source={1}
            style={
              Array [
                Object {
                  "height": "100%",
                  "resizeMode": "contain",
                },
              ]
            }
          />
        </View>
      </View>
      <View
        style={
          Array [
            Object {
              "alignItems": "center",
              "height": "18%",
              "justifyContent": "center",
              "marginBottom": "3%",
              "marginTop": "1%",
            },
          ]
        }
      >
        <Text
          style={
            Array [
              Object {
                "alignItems": "center",
                "justifyContent": "center",
                "paddingBottom": "3%",
                "paddingLeft": "3%",
                "paddingRight": "3%",
                "paddingTop": "3%",
              },
            ]
          }
        >
          Would You Like To Join Us? 
          texxt
        </Text>
        <View
          accessible={true}
          focusable={true}
          onClick={[Function]}
          onResponderGrant={[Function]}
          onResponderMove={[Function]}
          onResponderRelease={[Function]}
          onResponderTerminate={[Function]}
          onResponderTerminationRequest={[Function]}
          onStartShouldSetResponder={[Function]}
          style={
            Object {
              "alignItems": "center",
              "borderBottomWidth": 1,
              "justifyContent": "center",
              "opacity": 1,
            }
          }
        >
          <Text
            style={
              Array [
                Object {
                  "fontSize": 17,
                },
              ]
            }
          >
            Signup
          </Text>
        </View>
      </View>
    </View>
  </View>
</View>
`;
